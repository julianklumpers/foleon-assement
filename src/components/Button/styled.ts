import styled from "styled-components";

export const StyledButton = styled.button`
  border: none;
  background-color: ${({ theme }) => theme.colors.primary};
  width: 100%;
  color: #fff;
  font-size: 1.3rem;
  font-weight: 300;
  padding: 0.6em 1.2em;
  border-radius: 3px;
  cursor: pointer;
  white-space: nowrap;
`;
