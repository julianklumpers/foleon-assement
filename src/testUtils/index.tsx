import * as React from "react";
import { render } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { QueryClientProvider, QueryClient } from "react-query";
import { ThemeProvider } from "styled-components";
import { theme } from "../theme/theme";

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
    },
  },
});

const mountWithProviders = ({ children, wrappers, history }: any) => {
  const ComponentProviders = wrappers.reduceRight(
    (acc: any, provider: string) => {
      if (provider === "router") {
        return <Router history={history}>{acc}</Router>;
      }

      return acc;
    },
    children
  );

  return (
    <React.Suspense fallback="loading">
      <ThemeProvider theme={theme}>
        <QueryClientProvider client={queryClient}>
          {ComponentProviders}
        </QueryClientProvider>
      </ThemeProvider>
    </React.Suspense>
  );
};

const customRender = (
  ui: any,
  {
    wrappers = [],
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    ...options
  }: any = {}
) => {
  return {
    ...render(ui, {
      wrapper: (args) => mountWithProviders({ wrappers, history, ...args }),
      ...options,
    }),
    history,
    queryClient,
  };
};

export * from "@testing-library/react";
export { customRender as render, queryClient };
