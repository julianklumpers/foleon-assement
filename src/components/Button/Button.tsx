import * as React from "react";

import { StyledButton } from "./styled";

interface IButton extends React.HTMLAttributes<HTMLButtonElement> {
  type: "button" | "submit" | "reset";
}

export const Button: React.FC<IButton> = ({
  children,
  type = "button",
  ...rest
}) => {
  return (
    <StyledButton type={type} {...rest}>
      {children}
    </StyledButton>
  );
};
