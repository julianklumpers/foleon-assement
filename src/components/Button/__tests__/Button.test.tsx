import * as React from "react";
import { fireEvent, render, screen } from "@src/testUtils";
import { Button } from "../Button";

describe("Component: Button", () => {
  test("ui", () => {
    const { container } = render(<Button type="button" />);
    expect(container).toMatchSnapshot();
  });

  test("button with text", () => {
    const { container } = render(<Button type="button">TestTitle</Button>);

    expect(container).toHaveTextContent("TestTitle");
  });

  test("button is clickable", () => {
    const clickSpy = jest.fn();

    render(
      <Button type="button" onClick={clickSpy}>
        TestTitle
      </Button>
    );

    fireEvent.click(screen.getByText("TestTitle"));
    expect(clickSpy).toHaveBeenCalledTimes(1);
  });
});
