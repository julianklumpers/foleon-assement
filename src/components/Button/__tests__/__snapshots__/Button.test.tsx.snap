// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Button ui 1`] = `
.c0 {
  border: none;
  background-color: #36464e;
  width: 100%;
  color: #fff;
  font-size: 1.3rem;
  font-weight: 300;
  padding: 0.6em 1.2em;
  border-radius: 3px;
  cursor: pointer;
  white-space: nowrap;
}

<div>
  <button
    class="c0"
    type="button"
  />
</div>
`;
